src/goto-instrument/contracts/contracts.cpp

```diff
@@ -331,23 +381,11 @@ void code_contractst::check_apply_loop_contracts(
   havoc_gen.append_full_havoc_code(
     loop_head->source_location(), generated_code);
 
-  // Add the havocing code, but only check against the enclosing scope's
-  // write set if it was manually specified.
-  if(assigns_clause.is_nil())
-    insert_before_swap_and_advance(
-      goto_function.body,
-      loop_head,
-      add_pragma_disable_assigns_check(generated_code));
-  else
-    insert_before_swap_and_advance(
-      goto_function.body, loop_head, generated_code);
-
-  // ^^^ FIXME ^^^
-  // comment by delmasrd: I did not reactivate this behaviour
-  // because I think we always want to check the loop assignments against
-  // the surrounding clause
-
-  insert_before_swap_and_advance(goto_function.body, loop_head, generated_code);
+  // Insert the havocing code.
+  // We do not `add_pragma_disable_assigns_check`,
+  // so that the enclosing scope's assigns clause instrumentation
+  // would pick these up for inclusion (subset) checks.
+  goto_function.body.destructive_insert(loop_head, generated_code);
 
   // Generate: assume(invariant) just after havocing
   // We use a block scope to create a temporary assumption,
```

Comment:
at line 394 it is the second time this closure `invariant_expr()` gets called, is there a risk in executing the side effects of this closure twice ? or is it the very reason this is done using a closure ? Why not transform the invariant expression once and for all in the beginning ?

Reply:
resolved now; I was not aware of the fix that @tautschnig pointed out